### PR TITLE
add documentation for custom features

### PR DESCRIPTION
**Pull request by qjiang002**
Add doc and example json file for add custom features and analysis

### CODE REVIEW

File: docs/add_custom_features.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Add custom features for custom analysis
+
+If you want to perform custom analysis with your custom features that are not supported in the original task processors, and these features are only related to one particular system/dataset, you can define the custom features and analysis in the `metadata` section in the output JSON file. 
+
+Here is the template of the output JSON file with custom metadata.
```
Comment (neubig) [inline docs/add_custom_features.md:5]: This is nice, but it might be even more clear if you separately add one example of a discrete feature and one example of a continuous feature (and explicitly explain that you're doing so in the text).

File: docs/add_custom_features.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      "output-json-key": "model-output",
+      "custom-feature-name": "feature-value"
+    }
+  ]
+}
+```
+where
+* `feature-level` represents the fine-grained level of the analysis
+    * `example` for sentence-level analysis
+    * `span` for span-level analysis
```
Comment (neubig) [inline docs/add_custom_features.md:38]:
```suggestion
    * `span` for span-level analysis (e.g. in named entity recognition analysis)
    * `token` for token-level analysis (e.g. in named entity conditional text generation)
```

File: docs/add_custom_features.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+* `dtype` represents the data type of the feature
+    * `float` for continuous feature
+    * `string` for discrete feature
+* `cls_name` means the analysis type
+    * `BucketAnalysis` requires the following fields to be defined
+    ```
+    BucketAnalysis(
```
Comment (neubig) [inline docs/add_custom_features.md:45]: It might be a little bit confusing to suddenly have Python code here, as the expected input format is JSON. I'd suggest just having a bullet list. Similarly below.

**Review by neubig — COMMENTED** [review]
Thanks a lot, this is great! I have a few comments/suggestions.

**Comment by qjiang002** [comment]
In the second commit:
- Refine doc
- For ComboCountAnalysis, the json file only takes `list` but in the code it takes `tuple`. Add a type casting in analysis.py pass features as a tuple.

File: docs/add_custom_features.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
+# Add custom features for custom analysis
+
+If you want to perform custom analysis with your custom features that are not supported in the original task processors, and these features are only related to one particular system/dataset instead of the task itself, you can define the custom features and analysis in the `metadata` section in the output JSON file. 
+
+## Example of bucket analysis with discrete feature
```
Comment (neubig) [inline docs/add_custom_features.md:5]: Sorry to suggest changes multiple times, but could we give an example that has discrete features and continuous features in the same section/JSON? It seems that splitting it into two sections is a bit repetitive.

Otherwise this looks great, and I'm happy to approve after this is resolved!

File: data/system_outputs/text-classification-custom-feature-example.json [hunk 5]
```diff
@@ -0,0 +1,1035 @@
...
+          "description": "the subject in the example"
+        },
+        "adverb": {
+          "dtype": "string",
+          "description": "the adverb in the example"
+        },
+        "LL_delta": {
+          "dtype": "float",
+          "description": "log likelihood / confidence"
+        }
```
Comment (odashi) [inline data/system_outputs/text-classification-custom-feature-example.json:20]: The newest version of the repository requires to always specify "cls_name" for every feature.
```suggestion
        "verb": {
          "cls_name": "Value",
          "dtype": "string",
          "description": "the verb in the example"
        },
        "subject": {
          "cls_name": "Value",
          "dtype": "string",
          "description": "the subject in the example"
        },
        "adverb": {
          "cls_name": "Value",
          "dtype": "string",
          "description": "the adverb in the example"
        },
        "LL_delta": {
          "cls_name": "Value",
          "dtype": "float",
          "description": "log likelihood / confidence"
        }
```

File: docs/add_custom_features.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+Here is the output json format for bucket analysis with custom discrete features with `"dtype": "string"`, such as the subject of the sentence, etc.
+```
+{
+  "metadata": {
+    "custom_features": {
+      "feature-level": {
+        "custom-feature-name": {
+          "dtype": "string",
```
Comment (odashi) [inline docs/add_custom_features.md:13]: ditto

File: docs/add_custom_features.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+## Example of bucket analysis with continuous feature
+Here is the output json format for bucket analysis with custom continuous features with `"dtype": "float"`, such as the count of particular words, the output logits/probability, etc.
+```
+{
+  "metadata": {
+    "custom_features": {
+      "feature-level": {
+        "custom-feature-name": {
+          "dtype": "float",
```
Comment (odashi) [inline docs/add_custom_features.md:53]: ditto

File: data/system_outputs/text-classification-custom-feature-example.json (hunk 5 above)
Comment (qjiang002) [reply data/system_outputs/text-classification-custom-feature-example.json:20]: Thanks for the correction. What does "cls_name" mean for each feature? Is it the same as feature name or is it always 'Value'?

Comment (odashi) [reply data/system_outputs/text-classification-custom-feature-example.json:20]: "cls_name" is used to distinguish which type we have to map these objects to. If the feature is either number or string, it should be "Value".

**Review by pfliu-nlp — APPROVED** [review]
Thanks for the revision!

**Review by odashi — APPROVED** [review]

**Review by neubig — APPROVED** [review]
